### DISCUSSION TITLE
RP2350 testing

### DISCUSSION THREAD

**Discussion by znmeb** [tabemann/zeptoforth#131] [Ideas]
I am getting some boards with the new RP2350 micro-controller on them in the next couple of weeks. Speciifically I am getting

- [Pimoroni Pico Plus 2](https://shop.pimoroni.com/products/pimoroni-pico-plus-2?variant=42092668289107)
- [SparkFun Pro Micro - RP2350](https://www.sparkfun.com/products/24870)
- [Challenger+ RP2350 WiFi6/BLE5 ](https://ilabs.se/challenger-rp2350-wifi-ble/)
- [Challenger+ RP2350 BConnect](https://ilabs.se/challenger-rp2350-bconnect-datasheet/)

I should have the SparkFun boards Monday or Tuesday. I have no ETA yet on the others but they are coming from the UK and Sweden so it is at least a week.

RP2040 code should in theory run on these without modification but I will be rebuilding the Zeptoforth `uf2` files with the latest SDK before I do any testing. Are there any execution test suites for Zeptoforth I can run, assuming I get the code to come up?

**Comment by tabemann** [comment]
On Sat, Aug 10, 2024 at 7:43 PM M. Edward (Ed) Borasky <
***@***.***> wrote:

> I am getting some boards with the new RP2350 micro-controller on them in
> the next couple of weeks. Speciifically I am getting
>
>    - Pimoroni Pico Plus 2
>    <https://shop.pimoroni.com/products/pimoroni-pico-plus-2?variant=42092668289107>
>    - SparkFun Pro Micro - RP2350 <https://www.sparkfun.com/products/24870>
>    - Challenger+ RP2350 WiFi6/BLE5
>    <https://ilabs.se/challenger-rp2350-wifi-ble/>
>
> I should have the SparkFun boards Monday or Tuesday. I have no ETA yet on
> the others but they are coming from the UK and Sweden so it is at least a
> week.
>
> RP2040 code should in theory run on these without modification but I will
> be rebuilding the Zeptoforth uf2 files with the latest SDK before I do
> any testing. Are there any execution test suites for Zeptoforth I can run,
> assuming I get the code to come up?
>
The RP2040 and the RP2350 are supposedly in theory source-compatible if you
use the SDK or MicroPython/CircuitPython; they are not binary-compatible at
all. Note that this is very loosely so, though, because some things got
completely redesigned, such as the QSPI interface, and the boot system is
completely new. Also, most of the register base addresses and IRQ's changed
and some of the register offsets and bitfields changed, which further
breaks compatibility. Note that zeptoforth does not use the SDK; the only
outside dependencies zeptoforth has are the arm-none-eabi toolchain, make,
sh, Python 3, and if you want to use utils/codeload3.sh, pySerial.

Note though that I have already made good progress on a port of zeptoforth
to the RP2350. My work is in the rp2350 branch. Right now it has not been
tested; it has not even been compiled. I would welcome some alpha and beta
testing for it.

I ordered a few Raspberry Pi Pico 2's a couple days ago; they are
officially on backorder. Thanks for the pointer about the SparkFun boards,
but I see that they are already out of stock. I have ordered a couple
Pimoroni Pico Plus 2's but I will have to wait until their second shipment
comes in.

Travis

**Reply by znmeb** [reply]
Yeah, pretty much everything is pre-order / backordered, although the expensive iLabs boards may still be in stock. I don't have any plans to order any Raspberry Pi Pico 2 boards until the WH form factor is ready; most of my code requires WiFi and most of my testing requires soldered pins / breadboarding. 

Given that you have a port done, I'll test it out on the boards as they show up. All of the boards I ordered have PSRAM and off-chip flash, so I can test that too if it's in the code.

**Reply by znmeb** [reply]
Speaking of RP2040, I also have a couple of these: [TinyVison Pico-Ice](https://pico-ice.tinyvision.ai/). Another developer I know is building Forth cores in the FPGA but he is using Mecrisp in the RP2040 I think.

**Comment by tabemann** [comment]
Well my port isn't done per se - there are key parts that are missing, such
as support for the Always-On Domain and HSTX, and it is completely untested
so far.

Travis

**Reply by znmeb** [reply]
That;s OK - I can troubleshoot what's there and file issues. It won't be as fast as having a board connected to your USB port, but at least we aren't on different continents.

**Comment by tabemann** [comment]
On Sat, Aug 10, 2024, 23:10 M. Edward (Ed) Borasky ***@***.***>
wrote:

> That;s OK - I can troubleshoot what's there and file issues. It won't be
> as fast as having a board connected to your USB port, but at least we
> aren't on different continents.
>
Now the zeptoforth kernel and bootloader build successfully for the RP2350.
When one of your RP2350 boards arrive could you try flashing it with the
kernel, hooking up a USB-serial dongle to UART0 on GPIO's 0 and 1, and just
hitting enter in a terminal emulator and see if it responds with "ok"?

Thanks in advance,

Travis

> .
>

**Comment by znmeb** [comment]
OK ... it looks like the SparkFun boards will be here tomorrow. They're in Portland as of 21:21 PST8PDT and that usually means they get put on a delivery vehicle the next business day.

I'm checking the data sheet to see if the UART pins are on a connector or if I have to solder a connector on. If I have to solder it, I'll have to wait for the Pimoroni boards that already have a connector or find a makerspace that will let me wander in off the street and do it.

**Comment by znmeb** [comment]
I haven't found a pinout diagram but I did find a schematic - https://cdn.sparkfun.com/assets/6/2/1/0/8/SparkFun_ProMicro_RP2350.pdf

The only connector on it is a Qwiic connector - VCC, GND and SCL/SDA, which I'm assuming are serial clock and serial data. The UARTs are on the edges of the board in the normal places and I think I can pin the board to a breadboard with jumper wires at the corners and then connect UART0 to a Raspeberry Pi Debug probe, which has a UART-to-USB translator. If that doesn't work I'll have to solder headers on it.
